### PR TITLE
TS SDK - Event Queries

### PR DESCRIPTION
**Pull request by stella3d**
Add historical event queries

## New Methods

all new methods are on the rpc `Provider`.

* `getEventsByTransaction()`
* `getEventsByTransactionModule()`
* `getEventsByMoveEventStructName()`
* `getEventsBySender()`
* `getEventsByRecipient()`
* `getEventsByObject()`
* `getEventsByTimeRange()`

## Common Parameter Defaults

Parameters shared between many functions have defaults, to allow calling functions without specifying count or time range.

```typescript
async getEventsBySender(
    sender: SuiAddress,
    count: number = DEFAULT_RESULT_COUNT,
    startTime: number = DEFAULT_START_TIME,
    endTime: number = DEFAULT_END_TIME
  )
```

allows calling like:
```typescript
await rpc.getEventsBySender('0x0327d48f4c6e4cd8c7923b5940e3465e8198e9e2');
```
instead of
```typescript
await rpc.getEventsBySender('0x0327d48f4c6e4cd8c7923b5940e3465e8198e9e2', 1000, 0, Number.MAX_SAFE_INTEGER);
```

## API Design

The parameters for each method appear exactly as they do in the RPC method, with no abstraction. Initially i had an object that encapsulated a time range & count, but i backed it out in favor of a less opinionated design.

## Testing

here's the results of some manual testing.
### getEventsByTransaction()
image

### getEventsByObject()
image

### getEventsBySender()
image

### getEventsByRecipient()
image

### getEventsByTimeRange
image

### getEventsByModule
image

### getEventsByMoveEventStructName
image

### CODE REVIEW

File: sdk/typescript/src/types/events.ts [hunk 1]
```diff
@@ -12,7 +12,7 @@ export type MoveEvent = {
     transactionModule: string;
     sender: SuiAddress;
     type: string;
-    fields: { [key: string]: any; };
+    fields?: { [key: string]: any; };
```
Comment (stella3d) [inline sdk/typescript/src/types/events.ts:15]: `fields` should always have been optional, i think. Some transactions turned out to have events without fields, this change unblocked the type guard validation

File: explorer/client/src/components/events/eventDisplay.tsx [hunk 2]
```diff
@@ -67,7 +67,8 @@ function objectContent(label: string, id: ObjectId) {
     };
 }
 
-function fieldsContent(fields: { [key: string]: any }) {
+function fieldsContent(fields: { [key: string]: any } | undefined) {
+    if (!fields) return [];
```
Comment (stella3d) [inline explorer/client/src/components/events/eventDisplay.tsx:71]: this is related to my other comment about making `MoveEvent.fields` optional.

File: sdk/typescript/src/providers/provider.ts [hunk 3]
```diff
@@ -140,6 +145,97 @@ export abstract class Provider {
...
+  /**
+   * Get events emitted from within the specified Move module
+   * @param package_ Move package object ID
+   * @param module Move module name
+   * @param count max result count
+   * @param startTime start of time range
+   * @param endTime end of time range
+   */
+  abstract getEventsByModule(
+    package_: ObjectId,                   // 'package' is reserved word
```
Comment (666lcz) [inline sdk/typescript/src/providers/provider.ts:164]:
```suggestion
    packageId: ObjectId, 
```
You can get rid of this comment by using `packageId`

File: sdk/typescript/src/types/events.ts [hunk 4]
```diff
@@ -89,6 +89,12 @@ export type SuiEventEnvelope = {
...
 
+export type SuiEvents = SuiEventEnvelope[];
+
 export type SubscriptionId = number;
 
-export type SubscriptionEvent = { subscription: SubscriptionId, result: SuiEventEnvelope };
\ No newline at end of file
+export type SubscriptionEvent = { subscription: SubscriptionId, result: SuiEventEnvelope };
+
+export const DEFAULT_RESULT_COUNT = 1000;
```
Comment (666lcz) [inline sdk/typescript/src/types/events.ts:98]: Can we mirror the value in https://github.com/MystenLabs/sui/blob/e12f8c58ef7ba17205c4caf5ad2c350cbb01656c/crates/sui-json-rpc/src/api.rs#L27 ?
```suggestion
// mirrors the value defined in https://github.com/MystenLabs/sui/blob/e12f8c58ef7ba17205c4caf5ad2c350cbb01656c/crates/sui-json-rpc/src/api.rs#L27
export const EVENT_QUERY_MAX_LIMIT = 100;
```

**Review by 666lcz — CHANGES_REQUESTED** [review]
Thanks for adding the comprehensive test plan! Overall looks good, requesting changes for some renaming and documentation update(see inline comments). Could you also update the `Usage` section in https://github.com/MystenLabs/sui/tree/main/sdk/typescript#usage to include the usage of events related API?

**Pull request merged**